util/kvencoder/kv_encoder_test.go

```diff
@@ -0,0 +1,383 @@
+	tkExpect.MustExec("use test")
+
+	encoder, err := New(s.store, "test", nil, true)
+	c.Assert(err, IsNil)
+
+	schemaSQL := `create table t(
+		id varchar(20),
+		a char(10),
+		primary key(id))`
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+
+	sqls := []testCase{
+		{"insert into t values(1, 'test');", 0, 2},
+		{"insert into t(id, a) values(2, 'test')", 0, 2},
+		{"insert into t(id, a) values(3, 'test')", 0, 2},
+		{"insert into t values(1000000, 'test')", 0, 2},
+		{"insert into t(id, a) values(5, 'test')", 0, 2},
+		{"insert into t(id, a) values(4, 'test')", 0, 2},
+	}
+
+	s.runTestSQL(c, tk, tkExpect, encoder, sqls)
+}
+
+func insertData(tk, tkExpect *testkit.TestKit, sqls []string) {
+	for _, sql := range sqls {
+		tk.MustExec(sql)
+		tkExpect.MustExec(sql)
+	}
+}
+
+func (s *testKvEncoderSuite) TestUpdate(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	tkExpect := testkit.NewTestKit(c, s.storeExpect)
+	tk.MustExec("use test")
+	tkExpect.MustExec("use test")
+
+	encoder, err := New(s.store, "test", nil, false)
+	c.Assert(err, IsNil)
+
+	schemaSQL := `create table t(
+		id varchar(20),
+		a char(10),
+		primary key(id))`
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+	sqls := []string{
+		"insert into t values(1, 'test');",
+		"insert into t(id, a) values(2, 'test')",
+		"insert into t(id, a) values(3, 'test')",
+		"insert into t values(1000000, 'test')",
+		"insert into t(id, a) values(5, 'test')",
+		"insert into t(id, a) values(4, 'test')",
+	}
+	insertData(tk, tkExpect, sqls)
+
+	cases := []testCase{
+		{"update t set a = 'hello' where id = 1", 0, 1},
+		{"update t set a = 'hello' where id = 2", 0, 1},
+		{"update t set a = 'hello' where id = 10000000000000", 0, 0},
+		// remove old handle, and update primary key index, add new handle
+		{"update t set id = 10 where id = 1", 1, 3},
+	}
+	s.runTestSQL(c, tk, tkExpect, encoder, cases)
+}
+
+func (s *testKvEncoderSuite) TestDelete(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	tkExpect := testkit.NewTestKit(c, s.storeExpect)
+	tk.MustExec("use test")
+	tkExpect.MustExec("use test")
+
+	encoder, err := New(s.store, "test", nil, false)
+	c.Assert(err, IsNil)
+
+	schemaSQL := `create table t(
+		id varchar(20),
+		a char(10),
+		primary key(id))`
+	tk.MustExec(schemaSQL)
+	tkExpect.MustExec(schemaSQL)
+	sqls := []string{
+		"insert into t values(1, 'test');",
+		"insert into t(id, a) values(2, 'test')",
+		"insert into t(id, a) values(3, 'test')",
+		"insert into t values(1000000, 'test')",
+		"insert into t(id, a) values(5, 'test')",
+		"insert into t(id, a) values(4, 'test')",
+	}
+	insertData(tk, tkExpect, sqls)
+
+	cases := []testCase{
+		// remove handle and primary key
+		{"delete from t where id = 1", 2, 2},
+		{"delete from t where id = 1000000000", 0, 0},
+		{"delete from t where id = 2", 2, 2},
+	}
+	s.runTestSQL(c, tk, tkExpect, encoder, cases)
+}
+
+func (s *testKvEncoderSuite) TestRetryWithAllocator(c *C) {
```

Comment:
Does KvEncoder retry if we never commit a transaction?

Reply:
yes, qichao say there is some situation, the tools maybe retry, but want the api return the same kvs.